src/resolve/resolvectl.c

```diff
@@ -408,19 +408,49 @@ static int output_rr_packet(const void *d, size_t l, int ifindex) {
         return 0;
 }
 
+static int idna_candidate(const char *name, char **ret) {
+        _cleanup_free_ char *idnafied = NULL;
+        int r;
+
+        r = dns_name_apply_idna(name, &idnafied);
+        if (r < 0)
+                return log_error_errno(r, "Failed to apply IDNA to name '%s': %m", name);
+
+        if (r > 0 && !streq(name, idnafied)) {
+                *ret = TAKE_PTR(idnafied);
+                return true;
+        }
+
+        *ret = NULL;
+        return false;
+}
+
 static int resolve_record(sd_bus *bus, const char *name, uint16_t class, uint16_t type, bool warn_missing) {
         _cleanup_(sd_bus_message_unrefp) sd_bus_message *req = NULL, *reply = NULL;
         _cleanup_(sd_bus_error_free) sd_bus_error error = SD_BUS_ERROR_NULL;
+        _cleanup_free_ char *idnafied = NULL;
+        bool needs_authentication = false;
         unsigned n = 0;
         uint64_t flags;
-        int r;
         usec_t ts;
-        bool needs_authentication = false;
+        int r;
 
         assert(name);
 
         log_debug("Resolving %s %s %s (interface %s).", name, dns_class_to_string(class), dns_type_to_string(type), isempty(arg_ifname) ? "*" : arg_ifname);
 
+        if (dns_name_is_single_label(name))
+                log_notice("(Note that search domains are not appended when resolving raw record types. "
+                           "Please specify fully qualified domain names when resolving raw records, or remove --type= switch from invocation in order to request regular hostname resolution.)");
```

Comment:
"breaking"? I wouldn't say that, it#s just a log message we print after all, the operation goes through as it always went. But it is confusing, that's true.

I figure we could suppress the message for non A/AAAA requests, I think people wouldn't expect search domains to work for that. And then maybe suppress the message for special domain names such as _gateway, and localhost or so? For your two example lines this would already fix things.

Showing it only on failure I think is less than ideal since search domain issues are also about "masking" domains, not just about fialing lookups, if you understand what I mean. i.e. cases where you expect resolution in a local context, but then get internet scope, where it suceeds, too.

Reply:
> Hmm:
> 
> It seems like this warning is breaking many legitimate use cases. Should maybe the warning only be printed if the query fails?

I think it is wrong to disable search domain, when type of query is specified. Please avoid calling it 'raw' queries, noone else uses such name. Call it typed query or something. But _raw_ does not describe what it is supposed to do. I think _explicit record type_ would be more precise name for it.

All DNS lookup tools can enable search domains by special flag. Don't you think extra command line parameter should be used? It is bad practice to put multiple different actions into single function when programming. Function should do one thing, which is indicated by its name. If I specify type, please let me specify *only* type of query. Do not change other aspects, it would be quite confusing for anyone to use. One parameter should not modify multiple things. Especially not *-t* parameter, which is supported by all name lookup utilities without any similar additional changes to name processing. Including dig, host, kdig, drill, unbound-host. Please keep basic compatibility with them, when you offer quite similar tool. Thanks.